libs/ui/src/print_element.tsx

```diff
@@ -0,0 +1,127 @@
+import ReactDom from 'react-dom';
+import styled from 'styled-components';
+
+import { ElementWithCallback, PrintOptions } from '@votingworks/types';
+import { assert, getPrinter } from '@votingworks/utils';
+
+const PrintStyles = styled.div`
+  display: none;
+  background: #ffffff;
+  @media print {
+    display: block;
+  }
+`;
+
+// Wrapper that waits for all img elements within it to load before using
+// its callback
+function WrapperWithCallbackAfterImagesLoaded({
+  children,
+  onImagesLoaded,
+}: {
+  children: JSX.Element;
+  onImagesLoaded: () => void;
+}) {
+  const wrapperRef = useRef<HTMLDivElement>(null);
+  useEffect(() => {
+    assert(wrapperRef.current);
+    const imgElements = Array.from(wrapperRef.current.querySelectorAll('img'));
+    const unloadedImages = imgElements.filter(
+      (imgElement) => !imgElement.complete
+    );
+
+    let unloadedImageCount = unloadedImages.length;
+    if (unloadedImageCount === 0) {
+      onImagesLoaded();
+      return;
+    }
+
+    function onLoad() {
+      unloadedImageCount -= 1;
+      if (unloadedImageCount === 0) {
+        onImagesLoaded();
+      }
+    }
+
+    for (const unloadedImage of unloadedImages) {
+      unloadedImage.onload = onLoad;
+    }
+  }, [onImagesLoaded]);
+  return <div ref={wrapperRef}>{children}</div>;
+}
+
+// Render an element and print it. The function to render the element takes a
+// callback to indicate when the component has finished rendering and is ready
+// to be printed. This accommodates components that may want to do multiple
+// renders or post-processing before being ready to print.
+export async function printElementWhenReady(
+  elementWithOnReadyCallback: ElementWithCallback,
+  printOptions: PrintOptions
+): Promise<void> {
+  const printRoot = document.createElement('div');
+  printRoot.dataset['testid'] = 'print-root';
+  document.body.appendChild(printRoot);
+
+  return new Promise<void>((resolve) => {
+    let imagesLoaded = false;
+    let elementReady = false;
+    async function printAndTeardown() {
+      await getPrinter().print(printOptions);
+      ReactDom.unmountComponentAtNode(printRoot);
+      printRoot.remove();
+      resolve();
+    }
+    async function onImagesLoaded() {
+      imagesLoaded = true;
+      if (elementReady) {
+        await printAndTeardown();
+      }
+    }
+    async function onElementReady() {
+      elementReady = true;
+      if (imagesLoaded) {
+        await printAndTeardown();
+      }
+    }
+    ReactDom.render(
+      <PrintStyles>
+        <WrapperWithCallbackAfterImagesLoaded onImagesLoaded={onImagesLoaded}>
+          {elementWithOnReadyCallback(onElementReady)}
+        </WrapperWithCallbackAfterImagesLoaded>
+      </PrintStyles>,
+      printRoot
+    );
+  });
+}
+
+// Wrapper component to give a regular component an "onRendered"
+// callback prop that will get called after the first render of
+// the component finishes.
+function WrapperWithCallbackAfterFirstRender({
```

Comment:
I think keeping it is better than the two immediate alternatives. First alternative - pass it into `printElementWhenReady` with the images to load wrapper. Then we're double wrapping with that (without extra arguments) and that seems bad. Second alternative - remake the print logic within `printElement` this might be nice, but it's also nice to have a single entry point into printing. And in this case, our testing actually tests `printElementWhenReady` thoroughly. Even though - currently - it's not being used in the apps. It will be in election manager.

That was a bit of reason soup but, even though the wrapper isn't doing anything, I think it allows a better structure.

Reply:
Sorry, I wasn't totally clear. I meant that the logic of the wrapper is superceded by the logic of waiting for images to load.

Here's my reasoning:
- To check if images have loaded, we run a useEffect in the parent component of the element to be printed
- For that useEffect to run, the parent component must have rendered once
- For the parent component to have rendered once, all of its children must have rendered once
- Thus, if we execute the check if images have loaded, the element to be printed must have rendered once already

Does that seem right?

If so, I don't think `printElement` needs to use `WrapperWithCallbackAfterFirstRender`, since `printElementWhenReady` already guarantees that the element will be rendered once before printing.

Reply:
Since we made the image loading _not_ a wrapper component, I understand that to mean the existing wrapper is necessary.

Reply:
Yeah that sounds right